### PR TITLE
CHIP-0015: NFT Metadata Extension

### PR DESCRIPTION
**Pull request by BrandtH22**
Please note that this CHIP proposal is pending the next published release from schema.org (expected by end of May).
Feel free to message on the PR or me directly (bholmes22/@clydewallace22) with any questions/comments/concerns

--UPDATED-- This CHIP is in `Last Call` status as of `13 June 2023`

### CODE REVIEW

**Comment by danieljperry** [comment]
This CHIP is now a `DRAFT`. Please leave your reviews here.

**Comment by danieljperry** [comment]
We will be holding a public Zoom call to discuss this CHIP:
Time: May 16, 2023 at 4 PM PDT (11 PM UTC / 7 AM on the 17th in China)
Audience: Ecosystem developers and NFT creators
Zoom Link: See Discord
The call will be recorded and posted to Youtube afterward.

**Review by grigb — COMMENTED** [review]
The CHIP15 protocol enhances the interoperability, composability, and functional capacity of NFTs, by establishing a comprehensive and prescriptive structure for all classes of content.

By incorporating Schema.org's standards, CHIP15 encompasses a broad spectrum of novel applications which are presently unsupported and anticipates potential future requisites that remain unforeseen in the current digital landscape.

This protocol facilitates sophisticated integrations of intricate content types, including but not limited to; diverse file formats (documents, audio files, visual media, video files, 3D models, PDFs, etc.), various types of logical frameworks (legal contracts, game rules/mechanics, etc.), and an unlimited amount of assets and supplemental metadata associated with a single NFT.

By utilizing decentralized storage systems (IPFS, etc), any application can access the necessary data without the obligation of ownership or data storage. 

By using Schema.org's SHA256 attribute, every URL reference can be verified by client apps, and since the NFT metadata file is also verifiable, by using the same hash checksum standards, all the data for an NFT can be verified.

With the implementation of CHIP15, applications can access rich media types like videos and game assets simply by parsing a single JSON file containing the NFT's metadata.

_In this context, CHIP15 could be recognized as an essential component in the architecture of a decentralized digital future._



Schema.org's scope is vast. Here is the top level of the hierarchy:

Thing
    Action
    BioChemEntity
    CreativeWork
    Event
    Intangible
    MedicalEntity
    Organization
    Person
    Place
    Product
    Taxon

NFTs can be for anything, but many will use it for CreativeWork:

    AmpStory
    ArchiveComponent
    Article
    Atlas
    Blog
    Book
    Chapter
    Claim
    Clip
    Code
    Collection
    ComicStory
    Comment
    Conversation
    Course
    CreativeWorkSeason
    CreativeWorkSeries
    DataCatalog
    Dataset
    DefinedTermSet
    Diet
    DigitalDocument
    Drawing
    EducationalOccupationalCredential
    Episode
    ExercisePlan
    Game
    Guide
    HowTo
    HowToDirection
    HowToSection
    HowToStep
    HowToTip
    HyperToc
    HyperTocEntry
    LearningResource
    Legislation
    Manuscript
    Map
    MathSolver
    MediaObject
    MediaReviewItem
    Menu
    MenuSection
    Message
    Movie
    MusicComposition
    MusicPlaylist
    MusicRecording
    Painting
    Photograph
    Play
    Poster
    PublicationIssue
    PublicationVolume
    Quotation
    Review
    Sculpture
    Season
    SheetMusic
    ShortStory
    SoftwareApplication
    SoftwareSourceCode
    SpecialAnnouncement
    Statement
    TVSeason
    TVSeries
    Thesis
    VisualArtwork
    WebContent
    WebPage
    WebPageElement
    WebSite


Here's a complete list of Schema.org's data types:
https://schema.org/docs/full.html

**Comment by danieljperry** [comment]
Here is the discussion of this CHIP:
https://www.youtube.com/watch?v=gZ9V5GxT0GU

**Comment by yyolk** [comment]
**I fully support this additional CHIP. Leveraging schema.org is an obvious choice.**

* * * 

To anyone else that might have the question of where `@context` and `@type` may come from, and then stumbling onto the W3C recommendation for JSON-LD (JSON Link Data https://json-ld.org/). 

The following is my journey in discovering how `@context` will be used in the future. AFAICT this CHIP would be forward-compatible:

<details><summary>Here's one example where <code>@context</code> is a string, as it's used as I see CHIP-0015 recommending in some part, if I were to extrapolate a specific schema representing the NFT as a singular, whole:</summary>
https://www.w3.org/TR/json-ld11/#advanced-context-usage

</details>

<details>
<summary>Here's one where <code>@context</code> is an object:</summary>
<a href="https://www.w3.org/TR/json-ld11/#the-context">https://www.w3.org/TR/json-ld11/#the-context</a>

</details>

<details>
<summary>Another example shows that a <code>@type</code> is used in the 'leaf' node like shown in the CHIP-0015 examples.</summary>
<a href="https://www.w3.org/TR/json-ld11/#specifying-the-type">https://www.w3.org/TR/json-ld11/#specifying-the-type</a>

</details>

<details><summary>I saw JSON-LD gets very abstract, you can even alias keywords within the <code>@context</code>, so going that far can get pretty strange:</summary>
<a href="https://www.w3.org/TR/json-ld11/#aliasing-keywords">https://www.w3.org/TR/json-ld11/#aliasing-keywords</a>

</details>

I then wanted to see what would happen if I try to point `@context` to something *like* a `.jsonld`. (I went with the page for `VideoObject`, *which is not a `.jsonld`*

I ran from [the example](https://github.com/Chia-Network/chips/blob/6863bd7f6dd307b17dae64b0f83069cc90678e51/assets/chip-0015/examples/videoObject/mp4_w_integration1.json), the following: 

```json
{
        "@context": "https://schema.org/VideoObject",
        "@type": "VideoObject",
        "encodingFormat": "video/mp4",
        "thumbnail": {
            "@type": "ImageObject",
            "encodingFormat": "image/png",
            "name": "Pikachu Thumbnail",
            "url": "www.fakeurl.com/example1_thumbnail.png",
            "sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        }
}
```
The Validator gave this warning:
image


So coming full-circle it's appropriate to have `@context` set to `https://schema.org`. Assuming any `schema.org` validation would use the "local context of schema.org" this makes sense.


I believe there could be another CHIP in the future which could distill this down more, furthering compatibility with JSON-LD (if desirable). Although as a novice, I'm not well-versed enough to state whether it wouldn't be compatible as-is (pointing `@context` to a `schema.jsonld` at some URI). It might already be.

JSON-LD seems to be useful if you have a bunch of [custom](https://json-ld.org/contexts/person.jsonld) [schemas](https://json-ld.org/contexts/place.jsonld), that aren't yet introduced the schema.org where they may be slow on the uptake.

With the format, `CHIP-0015` being specified at the root of the metadata, it should be easy to iron any quirks out in the differences in the structure of `data`.

* * * 

tl;dr:

**"I fully support this additional CHIP. Leveraging schema.org is an obvious choice."**

File: CHIPs/chip-0015.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+ ```
+"data": {
+    "@context": "https://schema.org/",
+    "@type": "VideoObject",
+    "encodingFormat": "video/mp4",
+    "thumbnail": {
+        "@type": "ImageObject",
+        "encodingFormat": "image/png",
+        "name": "Example #1 Thumbnail",
+        "url": "www.fakeurl.com/example1_thumbnail.png",
```
Comment (yyolk) [inline CHIPs/chip-0015.md:202]: Sorry to be pedantic, should there be a missing `proto://` here? 

I think it would make sense to recommend falling back to `https` if theres no `proto`. It *should not* use `http` unless specified.

I brought this up during the zoom overview. I wanted to write it here for clarity.

I do recognize that after validating/parsing the data like the schema.org validator does - it falls back to `https://`. I also recognize that there's a section for [Verifying Data](https://github.com/Chia-Network/chips/blob/6863bd7f6dd307b17dae64b0f83069cc90678e51/CHIPs/chip-0015.md#verifying-data) which may cover this if I assume that's a strict requirement for usable URLs everywhere theres a URL. 

If the author of the `data: {...}` didn't include a `proto://` in the URL and it's like `ftp://` not `https://` that's on the author, similarly if it can only support `http` and no `proto` is included, that's on the author. Since the verification of this data with the validator.schema.org validator doesn't complain about the missing proto, and shows the intended URL with proto, I guess I could see `proto://` as optional when it's `https://` but I would not want to recommend it.

For posterity, here's how the data validates from [validator.schema.org](https://validator.schema.org):
image

File: assets/chip-0015/examples/videoObject/mp4_w_integration1.json [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    },
+    "data": {
+        "@context": "https://schema.org/",
+        "@type": "VideoObject",
+        "encodingFormat": "video/mp4",
+        "thumbnail": {
+            "@type": "ImageObject",
+            "encodingFormat": "image/png",
+            "name": "Pikachu Thumbnail",
+            "url": "www.fakeurl.com/example1_thumbnail.png",
```
Comment (yyolk) [inline assets/chip-0015/examples/videoObject/mp4_w_integration1.json:59]: missing `proto://` here?

File: assets/chip-0015/examples/videoObject/mp4_w_integration2.json [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    },
+    "data": {
+        "@context": "https://schema.org/",
+        "@type": "VideoObject",
+        "encodingFormat": "video/mp4",
+        "thumbnail": {
+            "@type": "ImageObject",
+            "encodingFormat": "image/png",
+            "name": "Pikachu Thumbnail",
+            "url": "www.fakeurl.com/pikachu_thumbnail.png",
```
Comment (yyolk) [inline assets/chip-0015/examples/videoObject/mp4_w_integration2.json:59]: missing `proto://` here?

File: assets/chip-0015/examples/videoObject/mp4_w_integration3.json [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    },
+    "data": {
+        "@context": "https://schema.org/",
+        "@type": "VideoObject",
+        "encodingFormat": "video/mp4",
+        "thumbnail": {
+            "@type": "ImageObject",
+            "encodingFormat": "image/png",
+            "name": "Pikachu Thumbnail",
+            "url": "www.fakeurl.com/pikachu_thumbnail.png",
```
Comment (yyolk) [inline assets/chip-0015/examples/videoObject/mp4_w_integration3.json:59]: missing `proto://` here?

**Review by yyolk — COMMENTED** [review]
Brought up some pedantic nit picks.

- Nit: missing `proto://` while not recommending a fallback to `https://` even though that's the behavior after "Verifying Data"
- Nit: `.iml` file checked in `untitled/untitled.iml`

File: untitled/untitled.iml
[the remark this answers is not in the document]
Comment (BrandtH22) [reply untitled/untitled.iml:1]: Thank you! This is now removed

File: assets/chip-0015/examples/videoObject/mp4_w_integration3.json (hunk 4 above)
Comment (BrandtH22) [reply assets/chip-0015/examples/videoObject/mp4_w_integration3.json]: added, thank you!

File: assets/chip-0015/examples/videoObject/mp4_w_integration2.json (hunk 3 above)
Comment (BrandtH22) [reply assets/chip-0015/examples/videoObject/mp4_w_integration2.json]: added, thank you!

File: CHIPs/chip-0015.md (hunk 1 above)
Comment (BrandtH22) [reply CHIPs/chip-0015.md]: added proto to the examples and updated url information to indicate it will default to https if not explicitly defined, thank you!

**Review by yyolk — COMMENTED** [review]
LGTM, thanks for accepting my suggestions!

**Comment by danieljperry** [comment]
This CHIP is now in `Review`. Brandt has updated it to include the latest spec, as well as many example files. Please leave your reviews here.

This CHIP will remain in `Review` status for a minimum of two weeks. After that point, if we have consensus from the community, it will be moved to `Last Call` before being finalized.

**Review by MarvinQuevedo — COMMENTED** [review]
This is an important evolution step to  make the NFT more scalable and compatible in different platforms and apps, good job! Congratulations!

**Comment by greimela** [comment]
Thanks for putting this together, @BrandtH22!

First of all, using schema.org for additional structured data is a great idea.
Why invent something new when there is such a rich standard already defined.
And I like the backwards-compatible way of how it is integrated, leveraging the `data` field from CHIP-0007.

From the perspective of [MintGarden](https://mintgarden.io), so an explorer and marketplace, this CHIP results in two things:
- More clarity of what NFTs represent and how they want to be displayed
- More work to support all the possible use cases

Apart from the thumbnail use case, which is a big improvement no questions asked, we will probably wait and see which types of data want to be used by creators.
The good thing is that we can always go back and upgrade existing NFTs based on the structured data that was originally provided but ignored.

I'm quite skeptical about the creator/collaborator information being part of the NFT metadata, I see that being the job of DIDs.

Also, this CHIP introduces a lot of use cases where links are embedded in the metadata, making them not as permanent as the ones directly added to the NFT, since they can't be upgraded if the hosting provider goes down.

**Comment by BrandtH22** [comment]
> Thanks for putting this together, @BrandtH22!
> 
> First of all, using schema.org for additional structured data is a great idea. Why invent something new when there is such a rich standard already defined. And I like the backwards-compatible way of how it is integrated, leveraging the `data` field from CHIP-0007.
> 
> From the perspective of [MintGarden](https://mintgarden.io), so an explorer and marketplace, this CHIP results in two things:
> 
>     * More clarity of what NFTs represent and how they want to be displayed
> 
>     * More work to support all the possible use cases
> 
> 
> Apart from the thumbnail use case, which is a big improvement no questions asked, we will probably wait and see which types of data want to be used by creators. The good thing is that we can always go back and upgrade existing NFTs based on the structured data that was originally provided but ignored.
> 
> I'm quite skeptical about the creator/collaborator information being part of the NFT metadata, I see that being the job of DIDs.
> 
> Also, this CHIP introduces a lot of use cases where links are embedded in the metadata, making them not as permanent as the ones directly added to the NFT, since they can't be upgraded if the hosting provider goes down.

Thank you for the detailed and thorough review!

Completely agreed that if we can start with thumbnails then see what additional fields are used by creators that will help greatly!
I do know a number of projects are already planning to use the isPartOf and hasPart features as well so I would recommend looking at those.

As for the creator/collaborator, you are correct that DIDs cover part of this but the reason to include the information in the extra data object is for all of those cases where the project DID does not include the information.

The main reason for this type of inclusion is that DIDs will innevitably be limited to creators/projects/organizations and will have difficulty bridging the gap to identify multiple different creators for a specific project or identifying creators of individual parts of an NFT. For example, imagine a musician who is releasing tracks; they would use their own creator DID to mint their NFT but what if they want to acknowledge the production company, the drum stem creator, or their videographer? Current DIDs, royalty split puzzles, and other coin contracts could fill this gap but that would lead to a large amount of additional blockchain data processing. Including the data in the NFTs metadata (potentially to include signatures from each individual creators DID that verifies their association with the project) would drastically reduce this parsing time.

While I believe this is another thing that we can wait to see how the creators in the space are using the fields and it is just another bit of data that schema.org can represent, I personally believe that having this data readily available will be beneficial for the long term.

**Comment by danieljperry** [comment]
Thanks for the reviews!
This CHIP is now in `Last Call`. If no changes are required for the next two weeks, it will become `Final`.

**Comment by danieljperry** [comment]
This CHIP is now `Final`. No further changes are allowed (other than errata).

**Review by danieljperry — APPROVED** [review]
lgtm
